_posts/2017-11-30-totw-119.md

```diff
@@ -0,0 +1,264 @@
+than your alias.
+
+This brittleness can be avoided if you do not add names to namespaces that you
+do not own (which includes the global namespace). By placing the aliases inside
+your own namespace, the unqualified lookup finds your alias first and never
+continues to search containing namespaces.
+
+More generally, the closer a declaration is to the point of use, the smaller is
+the set of scopes that can break your code. At the worst end of our example is
+`Quz`, which can be broken by anyone; `Bar` can only be broken by other code in
+`::example::util`, and a name that is declared and used inside an unnamed
+namespace cannot be broken by any other scope ([see below](#unnamed-namespaces)
+for an example).
+
+### Relative qualifiers
+
+A *using-declaration* of the form `using foo::Bar` seems innocuous, but it is in
+fact ambiguous. The problem is that it is safe to rely on the *existence* of
+names in a namespace, but it is not safe to rely on the *absence* of names.
+Consider this code:
+
+```c++
+namespace example {
+namespace util {
+
+using foo::Bar;
+```
+
+It is perhaps the author&rsquo;s intention to use the name `::foo::Bar`.
+However, this can break, because the code is relying on the existence of
+`::foo::Bar` _and also_ on the non-existence of namespaces `::example::foo` and
+`::example::util::foo`. This brittleness can be avoided by qualifying the used
+name fully: `using ::foo::Bar`.
+
+The only time that a relative name is unambiguous and cannot possibly be broken
+by outside declarations is if it refers to a name that is already inside your
+current namespace:
+
+```c++
+namespace example {
+namespace util {
+namespace internal {
+
+struct Params { /* ... */ };
+
+}  // namespace internal
+
+using internal::Params;  // OK, same as ::example::util::internal::Params
+```
+
+This follows the same logic that we discussed in the previous section.
+
+What if a name lives in a sibling namespace, such as `::example::tools::Thing`?
+You can say either `tools::Thing` or `::example::tools::Thing`. The fully
+qualified name is always correct, but it may also be appropriate to use the
+relative name. Use your own judgment.
+
+A cheap way to avoid a lot of these problems is not to use namespaces in your
+project that are the same as popular top-level namespaces (such as `util`); the
+Style Guide [recommends this practice]
+(http://engdoc/eng/doc/devguide/cpp/styleguide.shtml#Namespace_Names)
+explicitly.
+
+### Demo
+
+The following code shows examples of both failure modes.
+
+**helper.h:**
+
+```c++
+namespace bar {
+namespace foo {
+
+// ...
+
+}  // namespace foo
+}  // namespace bar
+```
+
+**some_feature.h:**
+
+```c++
+extern int f;
+```
+
+**Your code:**
+
+```c++
+#include "helper.h"
+#include "some_feature.h"
+
+namespace foo {
+void f();
+}  // namespace foo
+
+// Failure mode #1: Alias at a bad scope.
+// using foo::f;  // Error: redeclaration (because of "f" declared in some_feature.h)
```

Comment:
I feel like these should omit the prefix // for the actual code which is not a comment.